### ISSUE TITLE
proposal for naming conventions

### ISSUE THREAD

**Issue by MareoRaft** [jankapunkt/js-set-extension#4]
The following is a proposal for naming conventions of set operations (As always, there are no worries if you disagree, and everything is open to discussion).

I propose that the API/usage should match how the operation is phrased in mathematics in english.  So what you say determines the naming/syntax.  Also, I take into account that boolean functions are usually used in the context of if-statements.

Example
=======
What you say: "if A is a superset of B, ..."
What you type: `if (A.isSuperset(B)) {`

Proposed Naming Conventions
===========================

### containment

What you say: "if A is a proper superset of B, ..."
What you type: `if (A.isProperSuperset(B)) {`
note: After some nail biting, I settled on the word 'of' being builtin to functions.  We say "f of x" and we write "f(x)".  We don't write the of, but instead it is as-if the open-parenthesis is the of.

What you say: "if A is a proper subset of B, ..."
What you type: `if (A.isProperSubset(B)) {`

### equality

What you say: "if A is equal to B, ..." or "if A equals B, ..."
What you type: `if (A.equals(B)) {`

### unions

What you say: "The union of M"
What you type: `Set.union(M)`
ref: https://en.wikipedia.org/wiki/Union_(set_theory)#Arbitrary_unions
description: The arbitary union operator acts on any iterable (of sets) M and returns a single set whose elements consist of the elements of each set in the iterable.
note: Throws an error if you don't give it exactly 1 argument
note: Throws an error if the input is not an iterable of sets
note: The above two notes will be referred to as (*), to avoid repetition below.

What you say: "The union of A and B" or "A union B"
What you type: `A.union(B)`
ref: https://en.wikipedia.org/wiki/Union_(set_theory)#Union_of_two_sets
description: The small union operator is a binary operator which takes in two sets.
note: (*)
note: the function should be able to take in an iterable of length 0 or 1.
  * `Set.union([A])` should output A.
  * `Set.union([])` should output the empty set.

### intersections

What you say: "The intersection of M"
What you type: `Set.intersection(M)`
ref: https://en.wikipedia.org/wiki/Intersection_(set_theory)#Arbitrary_intersections
description: The arbitrary intersection operator acts on any iterable (of sets) and returns a single set whose elements consist of the elements that occur in every single set in the iterable.
note: (*)
note: the function should be able to take in an iterable of length 0 or 1.
  * `Set.intersection([A])` should output A.
  * `Set.intersection([])` should output the entire universe (but we don't have a global "universe" right now, so for now we should throw an error).

What you say: "The intersection of A and B" or "A intersect B"
What you type: `A.intersect(B)`
ref: https://en.wikipedia.org/wiki/Intersection_(set_theory)#Definition
description: The small intersection operator is a binary operator which takes in two sets.

### differences

What you say: "The set difference of A and B"
What you type: `Set.difference(A, B)`
note: no change

What you say: "A minus B"
What you type: `A.minus(B)`
description: This is the same as the difference operator, but with a different syntax.

What you say: "The symmetric difference of M"
What you type: `Set.symmetricDifference(M)`
ref: https://en.wikipedia.org/wiki/Symmetric_difference#n-ary_symmetric_difference
description: The repeated symmetric difference, as described in ref above.
note: This function *should* be able to take in an iterable of length 0 or 1.
  * `Set.symmetricDifference([A])` should output A.
  * `Set.symmetricDifference([])` should output the empty set.

What you say: "The symmetric difference of A and B"
What you type: `A.symmetricDifference(B)`
ref: https://en.wikipedia.org/wiki/Symmetric_difference
description: A triangle B.

### products

What you say: "The cartesian product of M"
What you type: `Set.cartesianProduct(M)`
description: Perform the n-ary cartesian product (aka the n-fold cartesian product) of all the sets in X
example:
  * `Set.cartesianProduct([A, B, C])` should output the set whose elements anything of the form (a, b, c) where a is in A, b is in B, and c is in C.
note: Note that `A.cartesianProduct(B).cartesianProduct(C)` is NOT the same as `Set.cartesianProduct([A, B, C])`.

What you say: "The cartesian product of A and B"
What you type: `A.cartesianProduct(B)`
ref: https://en.wikipedia.org/wiki/Cartesian_product
description: the binary operation version.

### power set

What you say: "The power set of A" or "The set of subsets of A"
What you type: `Set.powerSet(A)`
note: I have always heard people always say "the power set of A", but never "the power of A".  Hence, I believe truncating the function name to only "power" would cause confusion, since the term "power" can mean so many different things.

In conclusion
=============

Almost every set operation (with the exception of the difference operation, since it is not symmetric) has both a "binary operation version" (such as "A triangle B" or `A.triangle(B)`) and an "arbitrary operation version" (such as "triangle M" or `Set.triangle(M)`).  Our syntax reflects this.

**Comment by jankapunkt** [comment]
Hi @MareoRaft this is a great proposal as it extends the current library without breaking the existing API. I would create a new milestone and split the several features into single issues so we can track progress.

Nevertheless, I will keep this issue open as general discussion like guidelines, testing, integration etc.

**Comment by jankapunkt** [comment]
So this is how we will do this:

#### Branching

We will create a new branch for each issue, so we can create a pull request for each of the enhancements. This makes sure we will never break anything. It is important to work in small steps here, so the final release will be as stable as possible.

To lower the entry level, I will create for each issue a branch that has the current `master` branch as parent.

@MareoRaft I added you as collaborator, so you can directly push to the project and won't require to update your fork every time it's updated. If you prefer to work on your fork just let me know.

The issues are tracked automatically in the project: https://github.com/jankapunkt/js-set-extension/projects/1

#### Pull requests

We need for each pull request the CI build to pass. This also means, that tests have been updated to cover the new code, too.

#### Documentation

If there are new methods / features added to the source code, they need to be documented. We prefer to use [JsDoc](http://usejsdoc.org/) style comments and meta structure.

#### README

If there is information necessary to be accessible immediately to newcomers it should be put into the README.

**Comment by MareoRaft** [comment]
Thank you!  It seems that my proposal is a bit different than the current API though.  For example, we currently have

    function union (...args) {

which does not take in a single argument as proposed, but rather any amount of arguments.  Do you have a preference?  I'm happy with the spread syntax since apparently it works on any iterable (and so the user can type `Set.union(...myIterable)`.  The only downside is that it would not work if we wanted to backport the library to JavaScript 5.  But I'm assuming you don't want to do that.  For now I will leave the spread syntax as-is.

**Comment by jankapunkt** [comment]
Regarding the `union` method I forgot to clarify the following convention:

1. If the operation is binary (exactly two sets / variables are involved) it should be part of the instance's prototype.
2. If more sets / variables are involved (= arbitrary length) it should be a static member of the `Set` class.
3. The methods should always produce a new Set as result and should never mutate any of the involved parameters ([pure functions](https://en.wikipedia.org/wiki/Pure_function))

Does this makes sense to you?

Edit: I think this is also the way you already implemented the PR #13

**Comment by MareoRaft** [comment]
Yes, I was assuming that's what you wanted, but it's good to know explicitly.
